modules/standard/Sets.chpl

```diff
@@ -0,0 +1,700 @@
+    */
+    proc init=(const ref other: set(?t, ?)) {
+      this.eltType = t;
+      this.parSafe = other.parSafe;
+      this.complete();
+      for x in other {
+        //
+        // TODO: Not entirely sure if this inner copy is necessary, but we
+        // want to avoid potentially locking over and over.
+        //
+        var cpy = x;
+        _dom.add(cpy);
+      }
+    }
+
+    pragma "no doc"
+    inline proc deinit() {
+      clear();
+    }
+
+    pragma "no doc"
+    inline proc _enter() {
+      if parSafe then
+        _lock$.lock();
+    }
+
+    pragma "no doc"
+    inline proc _leave() {
+      if parSafe then
+        _lock$.unlock();
+    }
+ 
+    /*
+      Add a copy of the element `x` to this set. Does nothing if this set
+      already contains an element equal to the value of `x`.
+
+      :arg x: The element to add to this set.
+    */
+    proc add(in x: eltType) {
+      _enter();
+      const dbg = _dom.add(x);
+      _leave();
+    }
+
+    /*
+      Returns `true` if the given element is a member of this set, and `false`
+      otherwise.
+
+      :arg x: The element to test for membership.
+      :return: Whether or not the given element is a member of this set.
+      :rtype: `bool`
+    */
+    proc const contains(const ref x: eltType): bool {
+      _enter();
+      var result = _dom.contains(x);
+      _leave();
+      return result;
+    }
+
+    /*
+      Returns `true` if this set shares no elements in common with the set
+      `other`, and `false` otherwise.
+
+      :arg other: The set to compare against.
+      :return: Whether or not this set and `other` are disjoint.
+      :rtype: `bool`
+    */
+    proc const isDisjoint(const ref other: set(eltType, ?)): bool {
+      _enter();
+
+      var result = true;
+      for x in other do
+        // TODO: Can we insert a break in a for statement?
+        if !_dom.contains(x) then 
+          result = false;
+
+      _leave();
+
+      return result;
+    }
+
+    /*
+      Returns `true` if this set and `other` have at least one element in
+      common.
+
+      :arg other: The set to compare against.
+      :return: Whether or not this set and `other` intersect.
+      :rtype: `bool`
+    */
+    proc const isIntersecting(const ref other: set(eltType, ?)): bool {
+      return !isDisjoint(other);
+    }
+
+    /*
+      Remove the item from this set with a value equal to `x`.
+
+      .. warning::
+
+        Removing an element from this set may invalidate existing references
+        to the elements contained in this set.
+    
+      :arg x: The element to remove.
+
+      :throws IllegalArgumentError: If the list contains no such element.
+    */
+    proc remove(const ref x: eltType) throws {
+      _enter();
+
+      if !_dom.contains(x) {
+        _leave();
+        const msg = "No such element in set: " + x:string;
+        // TODO: Replace with more appropriate error?
+        throw new owned
```

Comment:
I'm not sure I like the fact that the set will throw when `parSafe=true`, because its possible another task has removed it. Why not just return true or false or something instead of throwing?

Reply:
Probably we should have a version that returns true/false.

Reply:
I have been thinking about this, and I think we should just change the return type of remove to:

```
proc remove(x: eltType): bool;
```

And along with this, get rid of the `discard` method.